language/move-compiler/src/typing/core.rs

```diff
@@ -1399,6 +1416,7 @@ fn check_type_argument_arity<F: FnOnce() -> String>(
 
 enum TVarCase {
     Single(String),
+    SingleButLast(String),
```

Comment:
This is super weird, so whatever we can do to remove this would be good. `Single` is supposed to be the case that allows references. And Base is any non-ref, non-tuple type. So these signify a sort of kind constraint. Shoe-horning this for `Fun` is I think just a result of not having a `Lambda` variant in `Type`?

Reply:
I think it's only a consequence of that now type arguments have different constraints. Alternatively, one could refactor this s.t. there is a different function to make the type params of function types, or pass the TVarCase for each param to make,